src/include/sof/ipc/common.h

```diff
@@ -187,7 +187,7 @@ ipc_cmd_hdr *mailbox_validate(void);
  *
  * @param hdr Points to the IPC command header.
  */
-void ipc_cmd(ipc_cmd_hdr *hdr);
+void ipc_cmd(ipc_cmd_hdr *_hdr);
```

Comment:
Just curious what is the warning here.

Reply:
That the signature in the header didnt match the implementation signature in the c files

Reply:
See https://github.com/thesofproject/sof/blob/main/src/ipc/ipc3/handler.c#L1513

Reply:
I disagree 100% with this check. I don't really care about this particular line but I think this check should be turned off.

The function signature is the same otherwise the code would not compile. That's how the compiler reports an actual signature mismatch:
```
sof/src/ipc/ipc3/handler.c:1513:6: error: conflicting types for 'ipc_cmd'
 1513 | void ipc_cmd(ipc_cmd_hdr *_hdr)
      |      ^~~~~~~
src/include/sof/ipc/common.h:191:6: note: previous declaration of 'ipc_cmd' was here
  191 | void ipc_cmd(ipc_cmd_hdr **hdr);
      |      ^~~~~~~
```
Parameter names in function declarations are totally optional and purely informative.

Having slightly different parameter names between the declaration and the definition is a very useful C feature. For instance:
- Slightly longer and more descriptive parameter names in the function declaration because there is no context.
- Slightly shorter parameter names in the function definition because there is all the context and repeating  long symbol names over and over in the function body makes  the code more verbose and less readable.

In this particular case neither `_hdr` nor `hdr` adds any information. The difference between the two serves a different purpose here: the underscore `_` has very specific and fairly clear meaning in the function code that makes no sense in the function declaration.

Again I don't really care about this particular line but it's a good example of how counter-productive this check is.

Reply:
This was a style level violation, the downside is some cppcheck checks that are low level informative/style are actually reporting serious issues such as memory leaks. We can disable this specific check, but style as a general thing should be on kept I think.

Reply:
> We can disable this specific check, but style as a general thing should be on kept I think.

I don't know about cppcheck in general, my comment was laser-focused on this particular violation.

Reply:
Lets fix the doxygen comment first and merge the fixes, we can continue cppcheck discussion afterwards.

Reply:
@marc-hb by "check" I mean this type of style violation. I.e. invoke cppcheck with --suppress="this_type_of_check"

@lgirdwood I will do it next week, I am OOO friday-monday

Reply:
> I disagree 100% with this check. I don't care about this particular line but I think this check should be turned off.
> 
> The function signature is the same otherwise the code would not compile. That's how the compiler reports an actual signature mismatch:
> 
> ```
> sof/src/ipc/ipc3/handler.c:1513:6: error: conflicting types for 'ipc_cmd'
>  1513 | void ipc_cmd(ipc_cmd_hdr *_hdr)
>       |      ^~~~~~~
> src/include/sof/ipc/common.h:191:6: note: previous declaration of 'ipc_cmd' was here
>   191 | void ipc_cmd(ipc_cmd_hdr **hdr);
>       |      ^~~~~~~
> ```
> 
> Parameter names in function declarations are totally optional and purely informative.
> 
> Having slightly different parameter names between the declaration and the definition is a very useful C feature. For instance:
> 
> * Slightly longer and more descriptive parameter names in the function declaration because there is no context.
> * Slightly shorter parameter names in the function definition because there is all the context and repeating  long symbol names over and over in the function body makes  the code more verbose and less readable.
> 
> In this particular case neither `_hdr` nor `hdr` adds any information. The difference between the two serves a different purpose here: the underscore `_` has a very specific and fairly clear meaning in the function code that makes no sense in the function declaration.
> 
> Again I don't care about this particular line but it's a good example of how counter-productive this check is.

No. Let's not be academic and rely on compilers.

it's a good practice to force variables to have the same names in .h and .c files. Usually this check raises more issues. We've cleaned this up for the kernel as well and the findings were priceless.

Reply:
> No. Let's not be academic

I'm the one advocating for more flexibility and fewer rules in this particular case.

> and rely on compilers.

You mean the C standard. Parameters in declarations are optional and useless for and ignored by _all_ compilers.

> the findings were priceless.

One example?

Reply:
> You mean the C standard. Parameters in declarations are optional and useless for and ignored by _all_ compilers.

Let's put it differently: Why on earth would a developer make a conscious attempt to name things differently in the header and the implementation? It's permitted by the standard but it's really stupid.

if you want examples, you can take a look at the cleanups I did for the kernel with 100+ of such changes. Inspecting the code and comparing declaration/definition helped flag additional issues with redundant inits and NULL pointer dereferences. It's really a telltale sign of code that needs to be reworked. Don't dismiss it on the grounds that the standard allows for it.

Reply:
> Why on earth would a developer make a conscious attempt to name things differently in the header and the implementation? It's permitted by the standard but it's really stupid.

I gave two examples above.

>  look at the cleanups I did for the kernel with 100+ of such changes.

I will try to find a few. How long ago?

> flag additional issues with redundant inits and NULL pointer dereferences

For now I really can't imagine how something ignored by all compilers can cause problems that bad.

> static analysis is always right.

I wish.

Reply:
Did you read my point: it's a sign that the code developers did not maintain consistency between different parts of the code, and a likely sign of more errors. So when we see such problems, we also look at the rest of the function. it's the same with initialized variables or duplicate initialization, usually that means error handling is not done right.

I don't see what's objectionable about this. It's not like the SOF code is perfect or even good.

Another example: it's perfectly legal to have shadowed variables *because the compiler knows what to do*. And yet this is a real problem leading to bugs.

Reply:
> Did you read my point: it's a sign that the code developers did not maintain consistency between different parts of the code, and a likely sign of more errors. So when we see such problems, we also look at the rest of the function.

I did miss that indirection sorry. So this is like looking for typos and then performing a second code review pass where the typo was found. I'm not surprised at all that a second pass of code review finds bugs considering how rushed the first pass can be but is this really the most efficient to find problems and does this find the most important issues? How about - for instance - spending crucially limited time and expertise raising the bar on the first code review pass instead?

There is a gazillion of analyzers and checks that go beyond compilers to try and correct C limitations and design issues. Turning them all at once usually drowns important issues in an ocean of minor or irrelevant issues or false positives. I don't even care so much about that particular check actually but I care a lot about finding the right balance of cppcheck settings so that @cujomalainey can finally enable it in CI, something he hasn't been able to do  because of the aforementioned "ocean" as he stated above, see Aug. 26th comments. For evidence that I care a 5 seconds look at https://github.com/thesofproject/sof/commits/main/.github is enough.

So will enabling this particular parameter name check in CI help detect some real problems? I can't imagine any for the moment but I'd love to be proven wrong.


>  because the compiler knows what to do.

I think there is a big difference between code and comments ignored by the compiler.  Both code and comments should be good (see the many  doxygen and comment fixes I made in #4509, #4441,  #4106, #3912, #2626,...) but there's a clear line between code and comments. Parameter names in declarations are not code, they're (important!) comments.

Reply:
cppcheck is still to verbose to be used for CI, specifically it does not work well with unions and our friends at AMD used unions. Coverity is the same, too verbose.

It's more of a maintainer tool that is run on a e.g. weekly basis - and before releases as we didn't do for 1.9... Oh well.

Reply:
I've seen the same warning - mismatched parameter name in declaration and definition without cppcheck. I think the compiler checks it itself (maybe enabled by some flags, or added in some versions), or maybe it was one of the tools, used during the kernel / SOF / Zephyr build. I wasn't using anything special for sure.